poetryenv

```diff
@@ -0,0 +1,141 @@
+
+    ./poetryenv poetryenvs/tf_min/ update
+    ./poetryenv poetryenvs/tf_min/ run task test
+"""
+import argparse
+import subprocess
+import sys
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Iterator, List
+
+import tomlkit
+
+_ROOT_DIR = Path(__file__).parent
+_BACKUP_DIR = _ROOT_DIR / "poetryenv_backup"
+_PYPROJECT_TOML = _ROOT_DIR / "pyproject.toml"
+_BACKUP_PYPROJECT_TOML = _BACKUP_DIR / _PYPROJECT_TOML.name
+_POETRY_LOCK = _ROOT_DIR / "poetry.lock"
+_VENV_DIR = _ROOT_DIR / ".venv"
+
+
+@contextmanager
+def shuffle(root: Path, env_dir: Path, derived: bool) -> Iterator[None]:
+    """
+    Shuffles files between the root dir, backup dir and environment dir.
+
+    1: Move the root file to the backup dir.
+    2: Move the corresponding file from the environment dir to the root.
+    3: Yield to allow processing of the environment file.
+    4: Move the environment file back to the environment dir.
+    5: Move the root file back to the root dir.
+
+    `derived` indicates whether the file is something Poetry produces.
+    If `derived` is `True` the files are allowed to be missing (Poetry might not have produced them
+    yet). Otherwise the files must exist.
+    If `derived` is `False` the file is not moved to- / from the backup directory, but is assumed to
+    be produced in another way.
+    """
+    backup_path = _BACKUP_DIR / root.name
+    env_path = env_dir / root.name
+
+    if derived:
+        if root.exists():
+            root.rename(backup_path)
+        if env_path.exists():
+            env_path.rename(root)
+    else:
+        root.rename(backup_path)
+
+    try:
+        yield
+    finally:
+        if derived:
+            if root.exists():
+                root.replace(env_path)
+            if backup_path.exists():
+                backup_path.replace(root)
+        else:
+            backup_path.replace(root)
+
+
+def read_toml(path: Path) -> tomlkit.TOMLDocument:
+    with open(path, "rt", encoding="utf-8") as f:
+        return tomlkit.load(f)
+
+
+def write_toml(data: tomlkit.TOMLDocument, path: Path) -> None:
+    with open(path, "wt", encoding="utf-8") as f:
+        tomlkit.dump(data, f)
+
+
+def override_toml(root: tomlkit.TOMLDocument, overrides: tomlkit.TOMLDocument) -> None:
+    for key, override_value in overrides.items():
+        if (key in root) and (hasattr(root[key], "items")):
+            override_toml(root[key], override_value)
+        else:
+            root[key] = override_value
+
+
+def update_pyproject_toml(overrides_path: Path) -> None:
+    pyproject = read_toml(_BACKUP_PYPROJECT_TOML)
+    overrides = read_toml(overrides_path)
+    override_toml(pyproject, overrides)
+    write_toml(pyproject, _PYPROJECT_TOML)
+
+
+def poetryenv(env_dir: Path, poetry_args: List[str]) -> int:
+    assert _PYPROJECT_TOML.is_file()
+    assert env_dir.is_dir()
+    _BACKUP_DIR.mkdir()
+
+    try:
+        with shuffle(_PYPROJECT_TOML, env_dir, derived=False), shuffle(
```

Comment:
(I think you can put the three shuffes all in parentheses, which should make them align better)

Reply:
> (I think you can put the three shuffes all in parentheses, which should make them align better)

It doesn't seem so. I can use a `contextlib.ExitStack` to make it a bit more readable, though.